nx/lib/nx/defn/grad.ex

```diff
@@ -446,6 +446,76 @@ defmodule Nx.Defn.Grad do
     [{x, reduce_g(x, opts, g)}]
   end
 
+  defp grad(:product, [x, opts], ans, g) do
+    # From https://people.maths.ox.ac.uk/gilesm/files/NA-08-01.pdf
+    # 2.2.4 we take that the reverse-mode gradient of a determinant
+    # is A_bar = C_bar . C . inv(A^T)
+
+    # We also know that the determinant of a diagonal matrix is the
+    # product of the diagonal items. This means that we can imagine
+    # our condensed product axes as being the diagonal of a matrix
+    # and apply the determinant grad on it.
+
+    # Also, if A is diagonal, A^T = A and inv(A) is as trivial
+    # as inverting the items. Also, C_bar and C are scalars.
+    # This means that a_bar = c_bar |> Nx.multiply(c) |> Nx.divide(a).
+
+    # First, let's condition our input matrix by substituting 0s with 1s
+    # Then, we move the reducing axes into a condensed innermost axis.
+
+    # After we have our matrix well conditioned, we can operate on the
+    # innermost axis as described above, and then we transpose the items
+    # back into place
+
+    axes =
```

Comment:
@jonatanklosko This one could really use your insight on some of the tensor shape  manipulations below

Reply:
perhaps the code can be simplified somewhere using `if opts[:keep_axes]`